src/app/util/odd-sized-integers.h

```diff
@@ -0,0 +1,215 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/util/attribute-storage-null-handling.h>
+#include <lib/support/TypeTraits.h>
+
+#include <limits>
+
+namespace chip {
+namespace app {
+
+namespace detail {
+template <int ByteSize, bool IsSigned>
+struct WorkingTypeMapper
+{
+};
+template <int ByteSize>
+struct WorkingTypeMapper<ByteSize, true>
+{
+    using WorkingType = int64_t;
+};
+template <>
+struct WorkingTypeMapper<3, true>
+{
+    using WorkingType = int32_t;
+};
+template <int ByteSize>
+struct WorkingTypeMapper<ByteSize, false>
+{
+    using WorkingType = uint64_t;
+};
+template <>
+struct WorkingTypeMapper<3, false>
+{
+    using WorkingType = uint32_t;
+};
+} // namespace detail
+
+template <int ByteSize, bool IsSigned>
+struct OddSizedInteger
+{
+    // WorkingType is the type we use to represent the value as an actual
+    // integer that we can do arithmetic, greater/less-than compares, etc on.
+    using WorkingType = typename detail::WorkingTypeMapper<ByteSize, IsSigned>::WorkingType;
+
+    // StorageType is the type "at rest" in the attribute store.  It's a
+    // native-endian byte buffer.
+    using StorageType = uint8_t[ByteSize];
+};
+
+template <int ByteSize, bool IsSigned>
+struct NumericAttributeTraits<OddSizedInteger<ByteSize, IsSigned>>
+{
+    using IntType = OddSizedInteger<ByteSize, IsSigned>;
+    // StorageType is the type "at rest" in the attribute store.  It's a
+    // native-endian byte buffer.
+    using StorageType = typename IntType::StorageType;
+    // WorkingType is the type we use to represent the value as an actual
+    // integer that we can do arithmetic, greater/less-than compares, etc on.
+    using WorkingType = typename IntType::WorkingType;
+
+    // BIGENDIAN_CPU to match how the attribute store works, because that's
+    // what where our data buffer is eventually ending up or coming from.
+#if BIGENDIAN_CPU
+    static constexpr int highIndex     = 0;
+    static constexpr int lowIndex      = ByteSize - 1;
+    static constexpr int lowerIndex    = 1;
+    static constexpr int raiseIndex    = -1;
+    static constexpr int pastLowIndex  = ByteSize;
+    static constexpr int pastHighIndex = -1;
+#else  // BIGENDIAN_CPU
+    static constexpr int highIndex     = ByteSize - 1;
+    static constexpr int lowIndex      = 0;
+    static constexpr int lowerIndex    = -1;
+    static constexpr int raiseIndex    = 1;
+    static constexpr int pastLowIndex  = -1;
+    static constexpr int pastHighIndex = ByteSize;
+#endif // BIGENDIAN_CPU
+
+    static constexpr WorkingType StorageToWorking(StorageType storageValue)
```

Comment:
Please unit test this

Reply:
it would be nice if parameters could also be changed in some way (e.g. having structs for the constexpr values) so that unit tests can validate both little and big endian implementations.

Reply:
@tlykkeberg-grundfos do you perhaps have time to write some tests for this?

Reply:
Fwiw, we can probably test the big vs little endian in separate test files by just defining BIGENDIAN_CPU around the header include.  But yes, if we rewrote that bit slightly we might be able to test in one file.

Reply:
Hi @bzbarsky-apple we will look into testing it yes. Is there some test framework you think we should do this in. I'm asking because I have not written any tests for the Matter SDK yet, other than the ones we do in .yaml files ;-)

Reply:
I would use some of the smaller tests as an example:

`src/lib/support/tests/TestSafeString.cpp`
`src/system/tests/TestTimeSource.cpp`

Reply:
For running tests (I am assuming a vscode docker image run)

```sh
source scripts/activate.sh
gn gen out
ninja -C out TestSafeString
./out/tests/TestSafeString
```

as an example. Tests are defined in the corresponding BUILD.gn and the test BUILD.gn is referenced in src BUILD.gn file:

https://github.com/project-chip/connectedhomeip/blob/master/src/lib/support/tests/BUILD.gn#L21
https://github.com/project-chip/connectedhomeip/blob/master/src/BUILD.gn#L85

Reply:
Perhaps this to be a two-stage thing? Extend the `TestCluster.yaml` with all the Integer24,40,48,56 bit stuff and then write some UnitTests to test the code using `nlunit-test`?

Reply:
That seems pretty reasonable, yes.  I also think we could land this with just the unit tests, once those exist, and do the yaml/zap/etc bits for TestCluster as a followup...

@tlykkeberg-grundfos Thank you!

Reply:
I'll try to look into the Unit Test part.

Reply:
Test  clusters are not unit tests. I am asking for unit tests

Reply:
@andy31415 Unit tests have been added.  Please take a look?